### PR TITLE
Shooter!

### PR DESCRIPTION
**Pull request by celskeggs**
The important parts:
- the creation of BangBangTalon as an abstraction to control a Talon with a Bang-Bang controller and its own internal encoder.
- implementation of a full shooter.

Not all of the roller control is implemented yet: only the shooter-relevant code has been added.

Plan for the shooter interface:
Two buttons: fire and warmup. While either of these is being held, the shooter will attempt to reach full speed. If fire is being held and the shooter is at full speed, the ball will be shot. The shooter will remain at speed until the ball has exited. When the shooter is told to start rolling, the final roller will run slightly backwards for a short period of time, to pull the ball away from the shooter. The shooter will not start until this has completed. The final roller will constantly and slowly run forwards to keep the ball pushed up against the flywheel, unless the flywheel is currently spinning, in which case it won't touch the ball.

I haven't been able to test this yet, since the CAN Talons aren't supported by the emulator. I'm going to work on that next.

Also, some deployment updates. Nothing fancy with that part.

### CODE REVIEW

File: ZukoAzula2016/src/org/team1540/zukoazula/BangBangTalon.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    private static FloatInput wrapForTests(String name, FloatInput normal) {
+        Cluck.publish(name + " Feedback", normal);
+        FloatCell testControl = new FloatCell();
+        Cluck.publish(name + " Override", testControl);
+        testControl.setWhen(0, FRC.inTestMode().onPress());
+        return FRC.inTestMode().toFloat(normal, testControl);
+    }
+
+    public void setup(BooleanInput activate) throws ExtendedMotorFailureException {
+        BooleanInput isBelowVelocity = velocity.atMost(ZukoAzula.mainTuning.getFloat(name + " Target Velocity", 1));
```
Comment (AidanSmith) [inline ZukoAzula2016/src/org/team1540/zukoazula/BangBangTalon.java]: `Target Speed` is already a value in the constructor that serves the same function. This is redundant.

File: ZukoAzula2016/src/org/team1540/zukoazula/Shooter.java [hunk 2]
```diff
@@ -12,40 +15,64 @@
...
+
+        TalonExtendedMotor talonRight = FRC.talonCAN(11);
+        talonRight.modGeneralConfig().configureReversed(false, true);
+        talonRight.modGeneralConfig().activateFollowerMode(talonLeft);
+
+        return talonLeft;
+    }
+
+    private static BooleanInput setupRollersForSpinup(BooleanInput spinup, BooleanInput fire) throws ExtendedMotorFailureException {
+        BooleanInput interlude = quickPauseTimer(100, spinup.onPress());
```
Comment (AidanSmith) [inline ZukoAzula2016/src/org/team1540/zukoazula/Shooter.java]: 100 and 500 from the last pauseTimer maybe should be tunable.

**Comment by AidanSmith** [comment]
Everything else makes sense, though it was a lot so I think @tytynguyen will have a good second opinion.

File: ZukoAzula2016/src/org/team1540/zukoazula/BangBangTalon.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        Cluck.publish(name + " At Speed", isUpToSpeed);
+    }
+
+    private static BooleanInput getThreeState(BooleanInput forceTrue, BooleanInput forceFalse) {
+        return new DerivedBooleanInput(forceTrue, forceFalse) {
+            private boolean state;
+
+            @Override
+            protected synchronized boolean apply() {
+                state |= forceTrue.get();
```
Comment (Tytynguyen) [inline ZukoAzula2016/src/org/team1540/zukoazula/BangBangTalon.java]: Can't this be expressed with more clarity and simpler with if statements?

File: ZukoAzula2016/src/org/team1540/zukoazula/Shooter.java [hunk 4]
```diff
@@ -12,40 +15,64 @@
...
-        bangBangControl(encoder, shooterTrigger, ZukoAzula.mainTuning.getFloat("Shooter Velocity", 1), 0, 1).send(shooterMotors);
-        Cluck.publish("Shooter Motors", shooterMotors);
+        rollerSpeed.send(FRC.talonSimpleCAN(8, FRC.MOTOR_FORWARD));
+
+        BangBangTalon shooter = new BangBangTalon(makeLinkedTalons(), "Shooter");
+
+        // passively roll to the wheel in teleop, while the shooter is stopped.
+        rollerSpeed.attach(rollerArb.addBehavior("Passive Roll", FRC.inTeleopMode().and(shooter.isStopped)), ZukoAzula.mainTuning.getFloat("Roller Passive Speed", 0.1f));
+
+        PauseTimer actuallyFiring = new PauseTimer(500);
```
Comment (Tytynguyen) [inline ZukoAzula2016/src/org/team1540/zukoazula/Shooter.java]: This timer doesn't allow us to shoot directly after picking up a ball, keeping the shooter running at all times.Do we need to have this timer?

File: ZukoAzula2016/src/org/team1540/zukoazula/Shooter.java (hunk 2 above)
Comment (Tytynguyen) [inline ZukoAzula2016/src/org/team1540/zukoazula/Shooter.java]: The usage of interlude could be clarified with a comment/name change.

File: ZukoAzula2016/src/org/team1540/zukoazula/Shooter.java (hunk 4 above)
Comment (celskeggs) [inline ZukoAzula2016/src/org/team1540/zukoazula/Shooter.java]: After looking over my code and running it in the emulator, I'm not certain that there's a problem. Obviously, there will have to be a higher-priority roller override for sucking in balls.
Is there another issue?

**Pull request merged**